This notebook is the blog / Chunk of my thoughts throughout the building process.

First I need to setup the pyproject.toml file, this file is a config file of what the project is and what resources we need.

Then I will make the basic structure of the whole project:

notebooks folder: annotate what I am doing throughout the project

scripts folder: data that I need for ML and sound analyze

src/my_kws folder:\
    __init__.py: The initial file\
    augment.py: data amplification\
    dataset.py: PyTorch Dataset\
    features.py: Extract characteristic\
    io.py: Audio I/O


Now here we have the code to only read audio file and output (mono information / sample rate / dtype)

In [ ]:
import soundfile as sf

def load_audio(
    path,
    target_sr,
    mono,
    target_length_s
):
    data, sr = sf.read(path)
    print(f"shape={data.shape}, sr={sr}, dtype={data.dtype}")
    return data

if __name__ == "__main__":
    x = load_audio(r"D:\Downloads\oboe_samples\oboe_samples\Ob-ord-A4-mf.wav", 16000, True, 1.0)

shape=(297437,), sr=44100, dtype=float64


Explain:
We define the function load_audio.\
path: path of the audio file\
target_sr: the target sample rate\
mono: if this is mono audio file\
target_length_s: the target length of file loaded

After we run the code, the result:\
shape=(297437,): this is the mono file, one dimentional.\
sr=44100: the sample rate is 44100Hz, we need to resample to 16000Hz.\
dtype=float64: we need to convert this to float32.\

In [ ]:
import numpy as np
import soundfile as sf
import torch
import torchaudio.functional as AF

def load_audio(
    path,
    target_sr,
    mono,
    target_length_s
):
    data, sr = sf.read(path)
    print(f"shape={data.shape}, sr={sr}, dtype={data.dtype}")

    #convert to float32
    data = data.astype(np.float32)

    #resample (if sr != target_sr)
    if sr != target_sr:
        waveform = torch.from_numpy(data).unsqueeze(0)
        waveform = AF.resample(waveform, sr, target_sr)
        data = waveform.squeeze(0).numpy()

    print(f"after resample: shape={data.shape}, sr={target_sr}, dtype={data.dtype}")
    return data

if __name__ == "__main__":
    x = load_audio(r"D:\Downloads\oboe_samples\oboe_samples\Ob-ord-A4-mf.wav", 16000, True, 1.0)

shape=(297437,), sr=44100, dtype=float64
after resample: shape=(107914,), sr=16000


Now we have two more steps: convert dtype and resample.

